# Hotel Facilito

## ¿Quién va a cancelar?

*Hotel Facilito* se está preparando para las vacaciones de verano, sin embargo, les preocupa que muchos de sus clientes cancelan de ultima hora, dejándoles con habitaciones vacías – sucede que algunos clientes simplemente no se presentan y, a pesar de que a veces se cobra una cuota de reservación, la gran mayoría de las ganancias se obtiene cuando los huéspedes pagan el resto al ocupar la habitación. 

Usando sus datos, les gustaría que les ayudaras a identificar a aquellos clientes que tienen más posibilidad de cancelar ya que les gustaría darles seguimiento para que si en caso de que requieran cancelar, se haga con la mayor antelación posible.

![](./header.png)

## Datos

*Hoteles Facilito* tiene dos sucursales, uno ubicado en la capital del estado, "City Hotel" y otro en una comunidad cercana a la costa, "Resort Hotel".

Los datos que te ha enviado están en formato CSV, en donde cada línea representa una reservación con los siguientes atributos:

  - `hotel`: Hotel en el que se hizo la reserva.
  - `is_canceled`: Indica si la reserva fue cancelada o no.
  - `lead_time`: Número de días que transcurrieron entre la fecha de ingreso de la reserva en el PMS y la fecha de llegada.
  - `arrival_date_year`: Año de la fecha de llegada.
  - `arrival_date_month`: Mes de la fecha de llegada con 12 categorías: "Enero" a "Diciembre".
  - `arrival_date_week_number`: Número de semana de la fecha de llegada.
  - `arrival_date_day_of_month`: Día del mes de la fecha de llegada.
  - `stays_in_weekend_nights`: Número de noches de fin de semana (sábado o domingo) que el huésped se quedó o reservó para quedarse en el hotel.
  - `stays_in_week_nights`: Número de noches de semana (lunes a viernes) que el huésped se quedó o reservó para quedarse en el hotel.
  - `adults`: Número de adultos.
  - `children`: Número de niños.
  - `babies`: Número de bebés.
  - `meal`: Tipo de comida reservada. Valor categórico.
  - `country`: País de origen. Las categorías se representan según ISO 3155–3:2013.
  - `market_segment`: Designación del segmento de mercado.
  - `distribution_channel`: Canal de distribución de la reserva.
  - `is_repeated_guest`: Valor que indica si el nombre de la reserva era de un huésped recurrente (1) o no (0).
  - `previous_cancellations`: Número de reservas anteriores que fueron canceladas por el cliente antes de la reserva actual.
  - `previous_bookings_not_canceled`: Número de reservas anteriores que no fueron canceladas por el cliente antes de la reserva actual.
  - `reserved_room_type`: Código del tipo de habitación reservada. El código se presenta en lugar de la designación por razones de anonimato.
  - `assigned_room_type`: Código del tipo de habitación asignada a la reserva. A veces, el tipo de habitación asignada difiere del tipo de habitación reservada debido a razones operativas del hotel (por ejemplo, sobreventa) o por solicitud del cliente. El código se presenta en lugar de la designación por razones de anonimato.
  - `booking_changes`: Número de cambios/modificaciones realizados a la reserva desde el momento en que se ingresó en el PMS hasta el momento del check-in o la cancelación.
  - `deposit_type`: Indicación de si el cliente hizo un depósito para garantizar la reserva.
  - `agent`: ID de la agencia de viajes que realizó la reserva.
  - `company`: ID de la empresa/entidad que realizó la reserva o es responsable de pagar la reserva.
  - `days_in_waiting_list`: Número de días que la reserva estuvo en lista de espera antes de ser confirmada al cliente.
  - `customer_type`: Tipo de reserva.
  - `adr`: Tarifa diaria promedio.
  - `required_car_parking_spaces`: Número de espacios de estacionamiento requeridos por el cliente.
  - `total_of_special_requests`: Número de solicitudes especiales realizadas por el cliente (por ejemplo, cama doble o piso alto).
  - `reservation_status`: Último estado de la reserva.
  - `reservation_status_date`: Fecha en que se estableció el último estado.
  - `name`: Nombre del cliente.
  - `email`: Correo electrónico del cliente.
  - `phone`: Teléfono del cliente.
  - `credit_card`: Últimos cuatro dígitos de la tarjeta de crédito del cliente.

Los datos que te ha enviado la compañía están en el archivo `hotel_bookings_training.csv`.

> En realidad los datos provienen de [este dataset de Kaggle](https://www.kaggle.com/datasets/mojtaba142/hotel-booking), y puedes consultar más sobre el origen de los datos [en esta publicación](https://www.sciencedirect.com/science/article/pii/S2352340918315191)

## ¿Qué métricas podemos medir? 

¿Accuracy? pero, y ¿si nos interesa otra cosa?

Nos interesa hallar a TODAS las personas que potencialmente pueden cancelar, y la verdad es que a nuestros clientes no les molesta mucho si les llamamos para confirmar su reserva.

## Ejercicio

1) La variable que tratamos de predecir es 'is_canceled' 
2) La respuesta que vamos a obtener es binaria - (Aprendizaje supervisado - Clasificación - Clasificación binaria (sí - no))
3) ¿Cuál es la  métrica que vamos a utilizar o qué metricas podemos medir ?   
   -  Una métrica es un número que nos permite medir el rendimiento del modelo comparando:   
    lo que el modelo predijo   
    vs.   
    lo que realmente ocurrió.   
    Por ejemplo, conceptualmente:   
    
    - Métrica	¿Qué mide? : 
    Accuracy	Qué porcentaje de predicciones fueron correctas   
    Precision	De las reservas que el modelo predijo como canceladas, cuántas realmente fueron canceladas   
    Recall	De todas las reservas que realmente fueron canceladas, cuántas detectó el modelo   
    F1-score	Combina Precision y Recall   
    ROC-AUC	Qué tan bien distingue el modelo entre las dos clases   
    Confusion Matrix	Muestra los aciertos y errores separados por clase   
    Log Loss	Evalúa la calidad de las probabilidades que produce el modelo   


In [12]:
import pandas as pd

In [13]:
hotel_bookings = pd.read_csv("hotel_bookings_training.csv")

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
hotel_bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119190 entries, 0 to 119189
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119190 non-null  object 
 1   is_canceled                     119190 non-null  int64  
 2   lead_time                       119190 non-null  int64  
 3   arrival_date_year               119190 non-null  int64  
 4   arrival_date_month              119190 non-null  object 
 5   arrival_date_week_number        119190 non-null  int64  
 6   arrival_date_day_of_month       119190 non-null  int64  
 7   stays_in_weekend_nights         119190 non-null  int64  
 8   stays_in_week_nights            119190 non-null  int64  
 9   adults                          119190 non-null  int64  
 10  children                        119186 non-null  float64
 11  babies                          119190 non-null  int64  
 12  meal            

### Sobre los datos personales...

Nos deshacemos de los datos personales – que además carecen de utilidad por ser únicos:

In [16]:
hotel_bookings.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,name,email,phone-number,credit_card
0,City Hotel,0,21,2015,September,36,1,0,4,2,...,Transient,105.0,0,0,Check-Out,2015-09-05,Timothy George,George_Timothy@mail.com,638-082-2623,************3627
1,City Hotel,0,20,2016,September,38,12,1,0,1,...,Transient,89.0,0,2,Check-Out,2016-09-13,Angela Carpenter,Angela.C@verizon.com,874-747-0095,************7934
2,City Hotel,0,2,2016,March,13,24,0,2,2,...,Transient,134.0,0,1,Check-Out,2016-03-26,Richard Scott,RScott80@verizon.com,311-116-4410,************1822
3,Resort Hotel,1,6,2016,April,17,21,0,1,2,...,Transient,73.0,0,0,Canceled,2016-04-18,Daniel Cooper,DanielCooper@yandex.com,375-297-1783,************7716
4,Resort Hotel,0,40,2015,August,34,20,2,3,2,...,Transient,176.8,1,1,Check-Out,2015-08-25,David Young,David_Young@gmail.com,910-698-3037,************4952


Hay información que no es útil, por ejemplo, número de tarjeta de crédito, correo, número de teléfono, etc. Por lo tanto, vamos a eliminar algunas columnas que no son útiles para nuestro análisis.

In [17]:
# Remove personal information of customers
hotel_bookings = hotel_bookings.drop(['name', 'email', 'phone-number', 'credit_card'], axis=1)

In [18]:
hotel_bookings.sample(10)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
12313,City Hotel,0,144,2016,July,31,25,1,1,2,...,No Deposit,7.0,NaN,0,Transient,81.08,0,0,Check-Out,2016-07-27
38122,Resort Hotel,0,265,2016,September,38,16,4,10,2,...,No Deposit,171.0,NaN,0,Transient,98.40,0,1,Check-Out,2016-09-30
40769,City Hotel,1,221,2016,August,34,17,0,2,2,...,No Deposit,9.0,NaN,0,Transient,85.50,0,0,Canceled,2016-06-19
117542,Resort Hotel,0,82,2017,May,19,8,1,1,1,...,No Deposit,241.0,NaN,0,Transient,58.52,0,1,Check-Out,2017-05-10
35966,City Hotel,1,100,2015,October,44,31,1,1,2,...,No Deposit,1.0,NaN,0,Transient-Party,62.00,0,0,Canceled,2015-10-19
64021,Resort Hotel,0,87,2016,October,44,23,2,5,2,...,No Deposit,171.0,NaN,0,Transient,70.00,0,0,Check-Out,2016-10-30
45211,City Hotel,1,202,2017,March,12,24,2,2,2,...,No Deposit,9.0,NaN,0,Transient,77.86,0,1,Canceled,2016-09-17
10416,Resort Hotel,1,383,2016,October,41,6,1,3,2,...,No Deposit,315.0,NaN,0,Transient-Party,48.00,0,0,Canceled,2016-03-04
71263,City Hotel,1,12,2016,January,5,26,7,20,2,...,No Deposit,9.0,NaN,0,Transient,76.55,0,0,Canceled,2016-01-20
84225,City Hotel,0,0,2016,February,9,24,0,1,2,...,No Deposit,9.0,NaN,0,Transient,97.00,0,1,Check-Out,2016-02-25


## EDA

In [10]:
from ydata_profiling import ProfileReport

KeyboardInterrupt: 

In [ ]:
profile = ProfileReport(hotel_bookings, title="Pandas Profiling Report")

In [ ]:
# profile.to_file("bookings_profile.html")
## file:///Users/antonio.feregrino/hub/cf-ml/hotel-facilito/bookings_profile.html

In [ ]:
#profile.to_file("bookings_profile.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

100%|██████████| 36/36 [00:05<00:00,  6.14it/s]


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

'Reservation_status' tiene alta correlación con 'is_canceled', ambas variables son lo mismo, si dejamos reservation_status , estariamos incurriendo  en lo que se lalma filtracion de datos   
¿Qué representa cada una?
En este dataset:
- is_canceled → indica si la reserva terminó cancelada: 0 o 1.   
- reservation_status → indica el estado final de la reserva: Canceled, Check-Out o No-Show.  

La razón más importante es:   
reservation_status contiene información que solo se conoce después de que el resultado de la reserva se ha producido.   

Eso es data leakage.  
El artículo original del dataset describe reservation_status como el último estado de la reserva y reservation_status_date como la fecha en que se estableció ese último estado. PubMed Central (PMC)   

## Data leakage

La variable `reservation_status` indica el estado de la reservación, esta es en realidad un reflejo de `is_canceled`, si la dejamos dentro de las variables de entrada estamos filtrando información al modelo. Este nos dará un resultado excelente, pero en realidad no nos servirá de nada en la vida real. Debemos sacarla del dataset junto con otras varaibles asociadas:

In [19]:
# Avoid data leakage
hotel_bookings = hotel_bookings.drop(['reservation_status', 'reservation_status_date'], axis=1)

## Separa la variable a predecir
Vamos a separar la variable objetivo del data set


In [20]:
is_canceled = hotel_bookings['is_canceled'].copy()
hotel_data = hotel_bookings.drop(['is_canceled'], axis=1)

'is_canceled'
Esta será tu variable objetivo, también llamada:    
- target    
- variable dependiente   
- variable a predecir   
- normalmente 'y'   

In [26]:
is_canceled.value_counts()

is_canceled
0    75039
1    44151
Name: count, dtype: int64

In [27]:
is_canceled 

0         0
1         0
2         0
3         1
4         0
         ..
119185    0
119186    0
119187    0
119188    0
119189    0
Name: is_canceled, Length: 119190, dtype: int64

In [28]:
hotel_data

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,...,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests
0,City Hotel,21,2015,September,36,1,0,4,2,0.0,...,A,2,No Deposit,9.0,NaN,0,Transient,105.0,0,0
1,City Hotel,20,2016,September,38,12,1,0,1,0.0,...,A,0,No Deposit,9.0,NaN,0,Transient,89.0,0,2
2,City Hotel,2,2016,March,13,24,0,2,2,0.0,...,A,0,No Deposit,9.0,NaN,0,Transient,134.0,0,1
3,Resort Hotel,6,2016,April,17,21,0,1,2,0.0,...,D,0,No Deposit,NaN,NaN,0,Transient,73.0,0,0
4,Resort Hotel,40,2015,August,34,20,2,3,2,0.0,...,D,0,No Deposit,250.0,NaN,0,Transient,176.8,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
119185,Resort Hotel,1,2016,June,26,21,0,1,1,0.0,...,A,0,No Deposit,104.0,NaN,0,Transient,79.0,0,0
119186,Resort Hotel,17,2016,October,45,30,1,0,2,0.0,...,A,0,No Deposit,NaN,346.0,0,Transient-Party,66.0,1,0
119187,City Hotel,1,2017,April,17,27,0,1,2,0.0,...,A,0,No Deposit,9.0,NaN,0,Transient,160.0,0,0
119188,Resort Hotel,10,2017,June,25,24,2,1,3,0.0,...,A,1,No Deposit,NaN,NaN,0,Transient,185.0,1,0


## Split dataset

El objetivo de dividir los datos es para que entrenen un mejor modelo, porq si mas tiempo tarde en dividir los datos, más informacion se va a ir filtrando en su desarrollo del modelo, ya que tenemos cerca de 120 000 datos si nos va a dar oportunidad de tener un conjunto de datos, de prueba y de validacion.

Y la idea de separar los datos antes de entrenar el modelo es precisamente para evitar que información del conjunto de prueba/validación influya en el entrenamiento.

In [29]:
# Calculate test and validation set size:
original_count = len(hotel_bookings)  #Cuenta el número de registros en el dataset 
training_size = 0.60 # 60% de los datos se utilizarán para entrenamiento
test_size = (1 - training_size) / 2    # 20% de los datos se utilizarán para prueba
training_count = int(original_count * training_size) # Calcula el número de registros para el conjunto de entrenamiento
test_count = int(original_count * test_size) # Calcula el número de registros para el conjunto de prueba
validation_count = original_count - training_count - test_count # Calcula el número de registros para el conjunto de validación

print(training_count, test_count, validation_count)

71514 23838 23838


TRAINING → desarrollar modelo    
VALIDATION → ajustar/elegir   
TEST → evaluación final   

             120 000     
                │
       ┌────────┼────────┐   
       ↓        ↓        ↓    
      60%      20%      20%   
    Training Validation  Test   
       │         │        │   
       ↓         ↓        ↓    
    aprende    ajusta   FINAL    

no garantiza por sí mismo que no exista leakage, todavía puedes tener leakage dentro de training.   
or ejemplo, si hotel_data todavía contiene:    
reservation_status    
reservation_status_date    

In [30]:
from sklearn.model_selection import train_test_split

train_x, rest_x, train_y, rest_y = train_test_split(hotel_data, is_canceled, train_size=training_count)
test_x, validate_x, test_y, validate_y = train_test_split(rest_x, rest_y, train_size=test_count)

print(len(train_x), len(test_x), len(validate_x))

71514 23838 23838


In [31]:
validate_x

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,...,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests
85265,City Hotel,133,2016,March,11,11,1,2,2,0.0,...,A,0,No Deposit,129.0,NaN,0,Transient-Party,84.33,0,1
57250,Resort Hotel,33,2016,October,44,27,1,3,2,0.0,...,D,0,No Deposit,240.0,NaN,0,Transient,80.50,1,0
33471,Resort Hotel,176,2017,March,13,27,2,5,2,0.0,...,A,0,No Deposit,240.0,NaN,0,Transient,43.03,0,0
12477,City Hotel,138,2016,September,40,25,2,1,3,0.0,...,D,0,No Deposit,9.0,NaN,0,Transient,162.00,0,1
86459,City Hotel,8,2016,August,33,10,2,4,2,0.0,...,D,0,No Deposit,14.0,NaN,0,Transient,155.70,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
55990,City Hotel,304,2015,August,34,17,1,1,2,0.0,...,A,0,Non Refund,1.0,NaN,0,Contract,62.00,0,0
7623,City Hotel,103,2017,August,33,17,2,5,3,0.0,...,D,1,No Deposit,22.0,NaN,0,Transient,109.65,0,1
36361,City Hotel,316,2016,May,22,26,0,3,2,0.0,...,B,0,No Deposit,30.0,NaN,0,Transient-Party,110.00,0,0
73498,City Hotel,19,2016,September,39,18,2,4,3,0.0,...,F,0,No Deposit,7.0,NaN,0,Transient,150.02,0,1


In [32]:
validate_x.value_counts('arrival_date_month')

arrival_date_month
August       2831
July         2423
May          2312
April        2225
October      2197
September    2167
June         2135
March        2016
February     1601
December     1375
November     1336
January      1220
Name: count, dtype: int64

## Variables a codificar - One-hot encoding

 - hotel, meal, distribution_channel, reserved_room_type, assigned_room_type, customer_type, deposit_type

Un modelo de machine learning no puede trabajar directamente con cadena , todo tiene que ser número, tratamos de convertir la columna a numero, con la tecnica on hot encoding, lo convierte en 2 columnas binarias, 0 y 1 

In [33]:
from sklearn.preprocessing import OneHotEncoder

In [34]:
one_hot_encoder = OneHotEncoder(sparse_output=False, handle_unknown="ignore")  #no quiero que mi salida sea dispersa, quiero que sea un array normal. handle_unknown="ignore" para que ignore las categorías desconocidas
#Estás creando un objeto que sabe cómo transformar categorías en columnas binarias.

In [35]:
one_hot_encoder.fit(train_x[['hotel']])


OneHotEncoder(handle_unknown='ignore', sparse_output=False)

In [37]:
one_hot_encoder.transform(train_x[['hotel']])

array([[0., 1.],
       [0., 1.],
       [1., 0.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

Número de columnas One-Hot = número de categorías aprendidas, salvo configuraciones especiales como drop.   

Y en tu caso hotel tiene dos categorías, por eso aparecen dos columnas binarias.

| hotel | City Hotel | Resort Hotel |   

| City Hotel | 1 | 0 |   
| Resort Hotel | 0 | 1 |   
| City Hotel | 1 | 0 |   

Probemos con otra variable distribution:chanel

In [38]:
one_hot_encoder.fit(train_x[['distribution_channel']])  

OneHotEncoder(handle_unknown='ignore', sparse_output=False)

In [40]:
hotel_bookings.value_counts('distribution_channel')

distribution_channel
TA/TO        97706
Direct       14616
Corporate     6670
GDS            193
Undefined        5
Name: count, dtype: int64

In [41]:
one_hot_encoder.transform(train_x[['distribution_channel']])

array([[0., 0., 0., 1., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0.],
       ...,
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0.]])

2 categorías → 2 columnas   
3 categorías → 3 columnas   
10 categorías → 10 columnas   
100 categorías → 100 columnas   

¿Qué hacemos con variables categóricas que tienen cientos o miles de categorías?   

Ahí ya entran estrategias como One-Hot Encoding, agrupación de categorías poco frecuentes, Target Encoding, Frequency Encoding, etc.

## Variables a binarizar

 - total_of_special_requests, required_car_parking_spaces, booking_changes, previous_bookings_not_canceled, previous_cancellations

In [42]:
from sklearn.preprocessing import Binarizer

Binarizer convierte una variable numérica en una variable de dos estados: 0 y 1, utilizando un umbral (threshold).   
valor ≤ threshold  → 0   
valor > threshold  → 1   

El Binarizer se utiliza cuando tienes una variable numérica y quieres transformarla en una variable binaria.   



Diferencia con lo que acabamos de estudiar   
Técnica	      |  Entrada	   |  Salida	          |    Objetivo   
```OneHotEncoder``` 	```Categórica```	    ```Varias columnas 0/1```	   ```Representar categorías```   
Binarizer	    Numérica	    Una variable 0/1	   Crear una condición mediante umbral   
StandardScaler	Numérica	    Numérica escalada	   Cambiar la escala   
MinMaxScaler	Numérica	    Numérica escalada	   Llevar a un rango   

In [ ]:
binarizer = Binarizer() 

In [44]:
b1= train_x[['total_of_special_requests']].sample(10)  #vemos los tipos de datos que tiene la columna total_of_special_requests en este caso valores 0,1,2,3,4
b1

,total_of_special_requests
41422,0
75577,0
83798,0
69457,2
78358,0
86078,1
79169,2
30354,0
118310,2
117793,1


In [45]:
binarizer.fit(b1[['total_of_special_requests']]) 

Binarizer()

In [46]:
binarizer.transform(b1[['total_of_special_requests']])

array([[0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1]], dtype=int64)

In [47]:
#Ahora vamos a comparar el resultado de la binarizacion y el los datos train_x para identificar como se comporta la binarizacion con los datos de entrenamiento.
b1['res']=binarizer.transform(b1[['total_of_special_requests']])
b1

,total_of_special_requests,res
41422,0,0
75577,0,0
83798,0,0
69457,2,1
78358,0,0
86078,1,1
79169,2,1
30354,0,0
118310,2,1
117793,1,1


El objetivo es crear una neuva variable `has_made_special_requests`

In [48]:

_ = train_x.copy()
binarizer.fit(_[['total_of_special_requests']])
_['has_made_special_requests'] = binarizer.transform(train_x[['total_of_special_requests']])

_[['total_of_special_requests', 'has_made_special_requests']].sample(10)

,total_of_special_requests,has_made_special_requests
62908,0,0
24057,1,1
114905,0,0
31001,0,0
63669,0,0
44886,1,1
13491,0,0
5321,2,1
86875,1,1
60953,0,0


## Variables a escalar

 - adr

Aquí escalar una variable significa cambiar su rango o escala numérica para que las variables tengan magnitudes comparables.

El escalado es especialmente relevante para algoritmos como:   

K-Nearest Neighbors (KNN)   
K-Means   
SVM   
regresión logística   
redes neuronales   
métodos basados en gradiente   

En cambio, algoritmos basados en árboles, como:   

Decision Tree   
Random Forest   
Gradient Boosting   

generalmente no necesitan escalado para funcionar correctamente, porque las decisiones se basan en puntos de corte de las variables.   

1. StandardScaler   
2. MinMaxScaler   
3. RobustScaler

In [56]:
train_x[['adr']]

,adr
70370,79.00
21637,64.00
19929,62.00
104871,126.00
6124,143.10
...,...
75421,69.21
31842,76.50
74255,109.80
36474,82.50


In [58]:
from sklearn.preprocessing import RobustScaler

In [59]:
scaler = RobustScaler()

In [61]:
_ = train_x.copy()
scaler.fit(_[['adr']])
_['adr_scaled'] = scaler.transform(train_x[['adr']])

_[['adr', 'adr_scaled']].sample(10)

,adr,adr_scaled
105962,67.50,-0.488455
39490,68.47,-0.471226
78482,95.00,0.000000
81384,175.56,1.430906
101167,112.67,0.313854
48344,208.66,2.018828
43572,189.00,1.669627
53533,95.00,0.000000
70790,67.80,-0.483126
97475,112.86,0.317229


## Variables a dejar como tal

 - stays_in_weekend_nights, stays_in_week_nights
 
 
 > El tratamiento de estas depende del modelo a usar

## Armando un pipeline de transformación

Un pipeline es una secuencia ordenada de pasos que se ejecutan automáticamente.

En tu problema del hotel, hasta ahora has estado haciendo cosas como:

datos   
     ↓   
eliminar columnas   
     ↓   
OneHotEncoder   
     ↓   
Binarizer    
     ↓   
Scaler   
     ↓   
modelo    

In [63]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import FeatureUnion, Pipeline

¿Qué es ColumnTransformer?   

Este es especialmente importante para tu dataset de Hotel Booking.   

Porque tienes diferentes tipos de columnas.   

Por ejemplo:   

Numéricas   
lead_time   
adr   
adults   
children   
total_of_special_requests   

Estas podrían necesitar: Scaler   

Categóricas   
hotel   
meal   
market_segment   
deposit_type   
customer_type   

Estas podrían necesitar: OneHotEncoder   


Entonces tenemos un problema:   

No queremos aplicar el mismo tratamiento a todas las columnas.   

Y aquí aparece:   

ColumnTransformer   

* ColumnTransformer Permite decir:   

"A estas columnas aplícales esta transformación y a estas otras columnas aplícales otra transformación."   

                 hotel_data   
                     │   
          ┌──────────┴──────────┐   
          ↓                     ↓   
    Columnas numéricas        Columnas categóricas   
          │                     │   
          ↓                     ↓   
    StandardScale           OneHotEncoder     
          │                     │   
          └──────────┬──────────┘   
                     ↓   
               datos transformados   
                     ↓   
                  MODELO   

* FeatureUnion sirve para combinar las salidas de diferentes transformaciones.    

                 DATOS   
                  │   
          ┌───────┴───────┐   
          ↓               ↓   
      Transformación A   Transformación B   
          ↓               ↓       
      Features A       Features B    
          │               │      
          └───────┬───────┘     
                  ↓     
           Features combinadas     

In [64]:
one_hot_encoding = ColumnTransformer([
    (
        'one_hot_encode',
        OneHotEncoder(sparse_output=False, handle_unknown="ignore"),
        [
            "hotel",
            "meal", 
            "distribution_channel", 
            "reserved_room_type", 
            "assigned_room_type", 
            "customer_type"
        ]
    )
])

In [71]:
one_hot_encoding.fit(train_x)

ColumnTransformer(transformers=[('one_hot_encode',
                                 OneHotEncoder(handle_unknown='ignore',
                                               sparse_output=False),
                                 ['hotel', 'meal', 'distribution_channel',
                                  'reserved_room_type', 'assigned_room_type',
                                  'customer_type'])])

In [73]:
one_hot_encoding.transform(train_x).shape

(71514, 38)

Esto es muy importante.   

Aquí el encoder aprende las categorías existentes en train_x para cada una de esas seis columnas.

In [72]:
one_hot_encoding.transform(train_x)

array([[0., 1., 0., ..., 0., 1., 0.],
       [0., 1., 0., ..., 0., 0., 1.],
       [1., 0., 1., ..., 0., 1., 0.],
       ...,
       [1., 0., 1., ..., 0., 1., 0.],
       [1., 0., 0., ..., 0., 1., 0.],
       [1., 0., 1., ..., 0., 1., 0.]])

Aquí ocurre la transformación.   

Ahora toma las categorías originales y las convierte en columnas binarias.   Cada categoría se convierte en una columna.



In [80]:
binarizer = ColumnTransformer([
    (
        'binarizer',
        Binarizer(),
        [
            "total_of_special_requests", 
            "required_car_parking_spaces", 
            "booking_changes", 
            "previous_bookings_not_canceled", 
            "previous_cancellations",
        ]
    )
])
    
one_hot_binarized = Pipeline([
    ("binarizer", binarizer),
    ("one_hot_encoder", OneHotEncoder(sparse_output=False, handle_unknown="ignore")),
])

In [81]:
scaler = ColumnTransformer([
    ("scaler", RobustScaler(), ["adr"])
])

In [82]:
passthrough = ColumnTransformer([
    (
        "passthrough",
        "passthrough",
        [
            "stays_in_week_nights",
            "stays_in_weekend_nights",
        ]
    )
])

In [83]:
feature_engineering_pipeline = pipe = Pipeline(
    [
        (
            "features",
            FeatureUnion(
                [
                    ("categorical", one_hot_encoding),
                    ("categorical_binarized", one_hot_binarized),
                    ("scaled", scaler),
                    ("pass", passthrough),
                ]
            ),
        )
    ]
)

In [84]:
feature_engineering_pipeline

Pipeline(steps=[('features',
                 FeatureUnion(transformer_list=[('categorical',
                                                 ColumnTransformer(transformers=[('one_hot_encode',
                                                                                  OneHotEncoder(handle_unknown='ignore',
                                                                                                sparse_output=False),
                                                                                  ['hotel',
                                                                                   'meal',
                                                                                   'distribution_channel',
                                                                                   'reserved_room_type',
                                                                                   'assigned_room_type',
                                                                                   'customer_type'])])),
                                                ('categorical_binarized',
                                                 Pipeline(steps=[('binarizer',
                                                                  ColumnTrans...
                                                                                                    'previous_bookings_not_canceled',
                                                                                                    'previous_cancellations'])])),
                                                                 ('one_hot_encoder',
                                                                  OneHotEncoder(handle_unknown='ignore',
                                                                                sparse_output=False))])),
                                                ('scaled',
                                                 ColumnTransformer(transformers=[('scaler',
                                                                                  RobustScaler(),
                                                                                  ['adr'])])),
                                                ('pass',
                                                 ColumnTransformer(transformers=[('passthrough',
                                                                                  'passthrough',
                                                                                  ['stays_in_week_nights',
                                                                                   'stays_in_weekend_nights'])]))]))])

In [86]:
feature_engineering_pipeline.fit(train_x) # Vamos a entrenar el pipeline con los datos de entrenamiento. Esto significa que el pipeline aprenderá cómo transformar los datos de entrenamiento, incluyendo cómo codificar las variables categóricas, cómo binarizar las variables y cómo escalar las variables numéricas.

Pipeline(steps=[('features',
                 FeatureUnion(transformer_list=[('categorical',
                                                 ColumnTransformer(transformers=[('one_hot_encode',
                                                                                  OneHotEncoder(handle_unknown='ignore',
                                                                                                sparse_output=False),
                                                                                  ['hotel',
                                                                                   'meal',
                                                                                   'distribution_channel',
                                                                                   'reserved_room_type',
                                                                                   'assigned_room_type',
                                                                                   'customer_type'])])),
                                                ('categorical_binarized',
                                                 Pipeline(steps=[('binarizer',
                                                                  ColumnTrans...
                                                                                                    'previous_bookings_not_canceled',
                                                                                                    'previous_cancellations'])])),
                                                                 ('one_hot_encoder',
                                                                  OneHotEncoder(handle_unknown='ignore',
                                                                                sparse_output=False))])),
                                                ('scaled',
                                                 ColumnTransformer(transformers=[('scaler',
                                                                                  RobustScaler(),
                                                                                  ['adr'])])),
                                                ('pass',
                                                 ColumnTransformer(transformers=[('passthrough',
                                                                                  'passthrough',
                                                                                  ['stays_in_week_nights',
                                                                                   'stays_in_weekend_nights'])]))]))])

In [90]:
feature_engineering_pipeline.transform(train_x).shape

(71514, 51)

In [91]:
transformed = feature_engineering_pipeline.fit_transform(train_x)
transformed.shape

(71514, 51)

In [92]:
transformed

array([[ 0.        ,  1.        ,  0.        , ..., -0.28419183,
         2.        ,  1.        ],
       [ 0.        ,  1.        ,  0.        , ..., -0.55062167,
         1.        ,  0.        ],
       [ 1.        ,  0.        ,  1.        , ..., -0.58614565,
         2.        ,  0.        ],
       ...,
       [ 1.        ,  0.        ,  1.        , ...,  0.26287744,
         5.        ,  2.        ],
       [ 1.        ,  0.        ,  0.        , ..., -0.22202487,
         1.        ,  2.        ],
       [ 1.        ,  0.        ,  1.        , ...,  0.21492007,
         1.        ,  2.        ]])

## Model training

In [93]:
# Get a fresh copy of the pipeline
from sklearn.base import clone

feature_transformer = clone(feature_engineering_pipeline)

features_train_x = feature_transformer.fit_transform(train_x)
features_validate_x = feature_transformer.transform(validate_x)

In [95]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC

model = RandomForestClassifier(n_estimators=100)

model.fit(features_train_x, train_y) #tengo que pasarle una matrix de datos ya transformados 

RandomForestClassifier()

* ¿Qué es Random Forest?   
Random Forest es un modelo formado por muchos árboles de decisión.   

* ¿Qué significa n_estimators=100?   
"Quiero que mi Random Forest tenga 100 árboles."   


* ¿Qué hace model.fit()?   
"Entrena el modelo utilizando mis variables de entrada y las respuestas conocidas."



                HOTEL BOOKING DATA   
                         │   
                         ↓   
                Eliminar información   
                  personal/leakage   
                         │
                         ↓
                  Separar X e y
                         │
                  ┌──────┴──────┐
                  ↓             ↓
                   X             y
             hotel_data      is_canceled
                  │
                  ↓
             Train/Test/
             Validation
                  │
                  ↓
             PREPROCESAMIENTO
                  │   
       ┌──────────┼──────────┐    
       ↓          ↓          ↓   
       OneHot        Binarizer      Scaler   
       │          │          │   
       └──────────┼──────────┘   
                  ↓
          features_train_x
                  │
                  ↓
          RANDOM FOREST
                  │
        100 árboles
                  │
                  ↓
             model.fit()
                  │
                  ↓
          MODELO ENTRENADO
                  │
                  ↓
          hacer predicciones
                  │
                  ↓
             EVALUACIÓN
                  │
       Accuracy / Precision /
       Recall / F1 / ROC-AUC

In [98]:
model.predict(features_train_x).shape

(71514,)

## Model validation

In [104]:
from sklearn.metrics import accuracy_score, recall_score

pred_y = model.predict(features_validate_x) #"Utiliza el modelo entrenado para hacer predicciones sobre los datos de validación."

print(accuracy_score(validate_y, pred_y))
print(recall_score(validate_y, pred_y))

0.8103448275862069
0.7215419501133787


* ¿Qué porcentaje de las predicciones fueron correctas?      

En tu resultado aparece:     

0.8088346337780015    


* 71.34 % de Recall.   

Pero aquí tenemos que entender qué significa ese 71.34 %, porque es diferente del Accuracy.   

El Recall responde a:   

De todas las reservas que realmente fueron canceladas, ¿cuántas consiguió detectar el modelo?    

 ¿Por qué el profesor está utilizando Recall?   
 Tu objetivo de negocio era:   

Detectar con anticipación las reservas con mayor posibilidad de cancelación para poder hacer seguimiento.   

Entonces puede ser especialmente importante no dejar escapar demasiadas cancelaciones reales.   

                 TRAIN
                   ↓
            model.fit(...)
                   ↓
            MODELO ENTRENADO
                   ↓
        features_validate_x
                   ↓
            model.predict(...)
                   ↓
                 pred_y
                   │
             ┌─────┴─────┐
             ↓           ↓
        validate_y      pred_y
          REAL        PREDICCIÓN
             │           │
             └─────┬─────┘
                   ↓
              COMPARACIÓN
                   ↓
       ┌───────────┴───────────┐
       ↓                       ↓
   Accuracy                  Recall
       80.88%                  71.34%

## Construcción del pipeline final

In [105]:
final_inference_pipeline = Pipeline([
    ("feature_engineering", clone(feature_engineering_pipeline)),
    ("model", RandomForestClassifier(n_estimators=100))
])

In [106]:
final_training_dataset = pd.concat([train_x, validate_x])
final_training_response = pd.concat([train_y, validate_y])

In [107]:
final_inference_pipeline.fit(final_training_dataset, final_training_response)

Pipeline(steps=[('feature_engineering',
                 Pipeline(steps=[('features',
                                  FeatureUnion(transformer_list=[('categorical',
                                                                  ColumnTransformer(transformers=[('one_hot_encode',
                                                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                                                 sparse_output=False),
                                                                                                   ['hotel',
                                                                                                    'meal',
                                                                                                    'distribution_channel',
                                                                                                    'reserved_room_type',
                                                                                                    'assigned_room_type',
                                                                                                    'customer_type'])])),
                                                                 ('categorical_binarized',
                                                                  P...
                                                                                                                     'previous_cancellations'])])),
                                                                                  ('one_hot_encoder',
                                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                                 sparse_output=False))])),
                                                                 ('scaled',
                                                                  ColumnTransformer(transformers=[('scaler',
                                                                                                   RobustScaler(),
                                                                                                   ['adr'])])),
                                                                 ('pass',
                                                                  ColumnTransformer(transformers=[('passthrough',
                                                                                                   'passthrough',
                                                                                                   ['stays_in_week_nights',
                                                                                                    'stays_in_weekend_nights'])]))]))])),
                ('model', RandomForestClassifier())])

## Model testing


In [ ]:
test_pred_y = final_inference_pipeline.predict(test_x) #Aquí está utilizando el modelo/pipeline final para hacer predicciones sobre test_x.

print(accuracy_score(test_pred_y, test_y))
print(recall_score(test_pred_y, test_y))

0.8149173588388288
0.7703394877903514


Por eso ahora estamos utilizando el conjunto TEST, que debe haber permanecido separado hasta este momento.   
* Accuracy ≈ 81.49 %   
Aproximadamente el 81.5 % de las predicciones fueron correctas en el conjunto TEST.   

* Recall ≈ 77.03 %   

En el contexto de tu problema:   

De todas las reservas que realmente fueron canceladas en el conjunto de test, el modelo identificó aproximadamente el 77 %.   

## Model persistence

Aquí está guardando el modelo entrenado en un archivo.
El archivo se llama:

inference_pipeline.joblib

In [112]:
from joblib import dump

dump(final_inference_pipeline, "inference_pipeline.joblib")

['inference_pipeline.joblib']

Esto es muy importante.   

Supongamos que tu modelo necesita:   

OneHotEncoder   
Binarizer   
Scaler   
RandomForest   

Si solamente guardaras:   

dump(model, ...)
   
estarías guardando únicamente el Random Forest.    

Pero después, cuando llegue una nueva reserva:   

hotel = City Hotel   
adr = 120   
meal = BB   
...

tendrías que recordar y ejecutar exactamente las mismas transformaciones antes de pasar los datos al Random Forest.   

Por eso es mucho más práctico guardar el pipeline completo:   

             final_inference_pipeline
                       │
          ┌────────────┴────────────┐
          ↓                         ↓
    PREPROCESAMIENTO              MODELO
          │                         │
    OneHot/Binarizer/Scaler      Random Forest
          │                         │
          └────────────┬────────────┘
                       ↓
                 PREDICCIÓN

      HOTEL DATA
                         ↓
                Separar X e y
                         ↓
              Train / Validation / Test
                         ↓
                PREPROCESAMIENTO
                         ↓
        ┌────────────────────────────────┐
        │ OneHotEncoder                  │
        │ Binarizer                      │
        │ Scaler                         │
        └────────────────┬───────────────┘
                         ↓
                  RANDOM FOREST
                         ↓
                     TRAINING
                         ↓
                    VALIDATION
                         ↓
                  ajustes finales
                         ↓
                       TEST
                         ↓
              Accuracy ≈ 81.49 %
              Recall ≈ 77.03 %
                         ↓
                 GUARDAR PIPELINE
                         ↓
          inference_pipeline.joblib

La idea final es muy buena: no estás guardando solamente el algoritmo; estás guardando todo el proceso necesario para transformar una nueva reserva y obtener su predicción.

## Y entonces... ¿a quién le hablamos?

* La idea es: ya tenemos un modelo entrenado, ahora llegan nuevos clientes/reservas y queremos predecir cuáles tienen mayor probabilidad de cancelar.

In [113]:
from joblib import load

ultimate_inference_pipeline = load("inference_pipeline.joblib")

In [115]:
new_customers = pd.read_csv("new_customers.csv")
new_customers.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,name,email,phone-number,credit_card
0,City Hotel,0,0,2016,March,13,22,0,1,2,...,NaN,0,Transient,99.0,0,1,Elizabeth Morton,ElizabethMorton@xfinity.com,218-662-6872,************5891
1,City Hotel,0,21,2016,March,13,23,0,3,2,...,NaN,0,Transient-Party,62.0,0,1,Virginia Ward,Virginia_Ward51@gmail.com,845-529-3632,************1071
2,City Hotel,0,418,2016,September,40,26,1,2,2,...,NaN,223,Transient-Party,107.0,0,0,Joseph Taylor,Joseph_T@protonmail.com,451-454-5767,************5326
3,City Hotel,0,58,2016,March,12,17,0,3,2,...,NaN,0,Transient,63.0,0,0,Sara Allen,Sara_Allen18@yandex.com,402-581-2687,************8597
4,Resort Hotel,0,130,2017,July,28,9,2,0,1,...,NaN,0,Transient-Party,178.0,0,1,John Black,Black.John47@yandex.com,541-901-5663,************9017


In [116]:
new_customers['will_cancel'] = ultimate_inference_pipeline.predict(new_customers)
new_customers[['proba_check_in', 'proba_cancel']] = ultimate_inference_pipeline.predict_proba(new_customers)

In [117]:
new_customers[['name', 'phone-number', 'will_cancel', 'proba_cancel']].sort_values(by='proba_cancel', ascending=False).head(20)

,name,phone-number,will_cancel,proba_cancel
90,Regina Pacheco,350-100-9605,1,1.000000
32,Renee Reed,970-325-8809,1,1.000000
63,Deanna Jenkins,229-507-3138,1,1.000000
46,Katelyn Jones,323-339-3265,1,1.000000
8,Carrie Tanner,392-436-1692,1,1.000000
91,Donald Alvarez,105-155-9939,1,1.000000
43,Joseph Lawson,166-493-3428,1,1.000000
40,Bryan Fitzpatrick,601-411-5278,1,0.995417
58,Cory Alexander,864-688-3246,1,0.993430
64,Daniel Ortiz,960-672-0720,1,0.993430


## De tarea...

 - Entrena un modelo con *data leakage* y ve si es sospechosamente bueno
 - ¿Qué pasa si el cliente no recibió la habitación que solicitó? (`reserved_room_type` vs `assigned_room_type`)
 - No consideramos las fechas en las que iba a quedarse, ¿y si las incluyes en tu modelo?
 - ¿Podemos usar validación cruzada? 
 - ¿Qué tal de la búsqueda de hiper parámetros?